### PR TITLE
London | ITP-May-2025 | Seddiq Azam | Module-Data-Groups | Sprint 2

### PR DESCRIPTION
**Pull request by sedazam**
You must title your PR like this:

REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME

For example,

London | May-2025 | Carol Owen | Quote-generator

Complete the task list below this message.
If your PR is rejected, check the task list.

-->

## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)
- [x] My changes meet the requirements of this task

## Changelist

Briefly explain your PR.

This pull request includes debugging fixes, implementation improvements, and testing updates across multiple files in Sprint-2 and Sprint-3. Key changes include resolving bugs in object property handling, enhancing functionality for utility functions, and adding comprehensive test cases to ensure correctness. Below is a categorized summary of the most important changes:

### Debugging Fixes:
* **`address.js`**: Fixed a bug where `address.houseNumber` was incorrectly accessed as `address[0]`. Updated the code to properly define the `address` object and access its properties. [[1]](diffhunk://#diff-a8799983ec7fc959a4415f0153eeb3fa5722e36e6a6575a7c02faa65388709eeR7-R12) [[2]](diffhunk://#diff-a8799983ec7fc959a4415f0153eeb3fa5722e36e6a6575a7c02faa65388709eeL15-R21)
* **`author.js`**: Corrected the iteration over object properties by replacing `for...of` with `Object.values()` to handle non-iterable objects. [[1]](diffhunk://#diff-72f04f4e9729a19324a96c28ab3f5c802b9f8902d87c09a584e2cb47b47102d8R3-R7) [[2]](diffhunk://#diff-72f04f4e9729a19324a96c28ab3f5c802b9f8902d87c09a584e2cb47b47102d8L14-R23)
* **`recipe.js`**: Adjusted the logging of recipe ingredients by iterating over the array and joining them with newlines for proper formatting. [[1]](diffhunk://#diff-b4769dd0af0822be5a9b369b23e2f462415142e9065eb1564c23a27ee056be5bR3-R7) [[2]](diffhunk://#diff-b4769dd0af0822be5a9b369b23e2f462415142e9065eb1564c23a27ee056be5bL15-R20)

### Implementation Enhancements:
* **`contains.js`**: Implemented the `contains` function to check for object properties with safety checks for invalid inputs.
* **`lookup.js`**: Added the `createLookup` function to generate a key-value mapping from an array of pairs.
* **`querystring.js`**: Fixed edge cases in `parseQueryString`, such as handling keys without values, and updated the test suite. [[1]](diffhunk://#diff-28939931c15b4851ce3d9a01779c07af4306743c48e45062f6b90f85b2e2f840L8-R24) [[2]](diffhunk://#diff-11da742f3ed0d8c40988d908b75d0ed028c9241c4810c53e48cab495697f628dL6-R16)
* **`tally.js`**: Implemented the `tally` function to count occurrences of items in an array, with error handling for invalid inputs. [[1]](diffhunk://#diff-54ea4b941ba199f9c06188c264fc4c1979bd8ea5dc1666a77933bbd41e8c752cL1-R17) [[2]](diffhunk://#diff-611b0e3310d029b1402d0d2a61671a61b082cb569e54ed6c568cd152639dcda5R22-R50)
* **`invert.js`**: Corrected the `invert` function to dynamically swap object keys and values and added explanatory comments.

### Testing Improvements:
* **`contains.test.js`**: Added comprehensive tests for the `contains` function, covering valid and invalid inputs, empty objects, and edge cases.
* **`lookup.test.js`**: Added a test for `createLookup` to validate the output structure and correctness.
* **`tally.test.js`**: Expanded test coverage for the `tally` function, including cases for empty arrays, duplicate items, and invalid inputs.

### Stretch Goals:
* **`count-words.js`**: Implemented a function to count word occurrences in a string, including handling empty or whitespace-only strings.
* **`mode.js`**: Refactored `calculateMode` by splitting it into smaller functions for frequency calculation and finding the most frequent value. [[1]](diffhunk://#diff-a84758363d9dbe7eb4b4d9608ecb0fe59dea9977e9667946d7ee14c8ee123e13L11-R30) [[2]](diffhunk://#diff-a84758363d9dbe7eb4b4d9608ecb0fe59dea9977e9667946d7ee14c8ee123e13R40-R46)
* **`till.js`**: Fixed and added an alternative implementation for `totalTill` to correctly calculate total amounts in pounds, with detailed explanations and tests. [[1]](diffhunk://#diff-87b9872087b9b54514b29a49fe6de586e3b679425b512d79c528942d360e5d37L11-R22) [[2]](diffhunk://#diff-87b9872087b9b54514b29a49fe6de586e3b679425b512d79c528942d360e5d37L23-R64)

### Miscellaneous:
* **`package.json`**: Updated `main` in Sprint-2 and fixed a typo in the `test` script in Sprint-3. [[1]](diffhunk://#diff-678fb9a60609bd580230045fbd27c9c6294d4ecf4247c8f751fe013df4333b0eL5-R11) [[2]](diffhunk://#diff-ad0af50eed7564e4d672970e6826162c24bd4e7832c4092fdb1e7084ce466629L7-R7)<!--

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-2/implement/contains.js [hunk 1]
```diff
@@ -1,3 +1,16 @@
-function contains() {}
+function contains(obj, property) {
+  // Safety check: handle invalid inputs
+  if (
+    typeof obj !== "object" ||
+    obj === null ||
+    obj === undefined ||
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js:6]: If obj is undefined, the condition at line 4 would have been true.

File: Sprint-2/implement/lookup.test.js [hunk 2]
```diff
@@ -33,3 +31,17 @@ It should return:
    'CA': 'CAD'
  }
 */
+Test(
```
Comment (cjyuan) [inline Sprint-2/implement/lookup.test.js:34]: Does "Test" work the same as "test"?

File: Sprint-2/implement/querystring.js [hunk 3]
```diff
@@ -5,12 +5,21 @@ function parseQueryString(queryString) {
...
+    const pair = keyValuePairs[i];
+    const firstEqualIndex = pair.indexOf("=");
+
+    if (firstEqualIndex === -1) {
+      // If there is no equal sign, treat the whole pair as a key with an empty value
+      queryParams[pair] = "";
+    } else {
+      const key = pair.slice(0, firstEqualIndex);
+      const value = pair.slice(firstEqualIndex + 1);
+      queryParams[key] = value;
```
Comment (cjyuan) [inline Sprint-2/implement/querystring.js:18]: Please note that in real querystring, both `key` and `value` are ***percent-encoded*** or ***URL encoded*** in the URL. For example, the string "5%" will be encoded as "5%25". So to get the actual value of "5%25" (whether it is a key or value in the querystring), you should call a function to decode it.
May I suggest looking up any of these terms, and "How to decode URL encoded string in JS"?

File: Sprint-2/implement/tally.test.js [hunk 4]
```diff
@@ -19,16 +19,32 @@ const tally = require("./tally.js");
 // Given a function called tally
 // When passed an array of items
 // Then it should return an object containing the count for each unique item
+test("tally should return an object with counts for each unique item", () => {
+  const input = ["a", "b", "a", "c"];
+  const expectedOutput = { a: 2, b: 1, c: 1 };
+  expect(tally(input)).toEqual(expectedOutput);
+});
```
Comment (cjyuan) [inline Sprint-2/implement/tally.test.js:26]: Can you check if your function returns what you expect from the following function call?
```
  tally(["toString", "toString", "toString"]); 
```

File: Sprint-2/stretch/count-words.js [hunk 5]
```diff
@@ -26,3 +26,33 @@
...
+  for (const eachWord of words) {
+    if (eachWord in counts) {
+      counts[eachWord]++;
+    } else {
+      counts[eachWord] = 1;
+    }
+  }
+
+  return counts;
+};
```
Comment (cjyuan) [inline Sprint-2/stretch/count-words.js:46]: Can you check if your function returns what you expect in the following function calls?

```
countWords("constructor constructor");
countWords("          Hello       hello   ");
```

File: Sprint-2/stretch/till.js [hunk 6]
```diff
@@ -8,7 +8,18 @@ function totalTill(till) {
   let total = 0;
 
   for (const [coin, quantity] of Object.entries(till)) {
-    total += coin * quantity;
+    total += parseInt(coin) * quantity;
+  }
+
+  return `£${total / 100}`;
```
Comment (cjyuan) [inline Sprint-2/stretch/till.js:14]: Could consider using `.toFixed()` to ensure 2 decimal places.

File: Sprint-2/stretch/till.js [hunk 7]
```diff
@@ -8,7 +8,18 @@ function totalTill(till) {
...
+}
+
+function totalTillAlternative(till) {
+  let total = 0;
+
+  for (const [coin, quantity] of Object.entries(till)) {
+    // Convert coin to a number and multiply by quantity
+    total += Number(coin.slice(0, -1)) * quantity; // Remove 'p' from coin string
```
Comment (cjyuan) [inline Sprint-2/stretch/till.js:22]: Another approach would be to prepare a lookup table as `const table = { "1p": 1, "5p': 5, ... }`, then lookup the value of the coin as `table[coin]`.

File: Sprint-2/implement/contains.js (hunk 1 above)
Comment (sedazam) [reply Sprint-2/implement/contains.js]: Yes, `typeof obj !== "object"` already covers the case where `obj === undefined`; therefore, the `obj === undefined` check is redundant and has been removed.

File: Sprint-2/implement/lookup.test.js (hunk 2 above)
Comment (sedazam) [reply Sprint-2/implement/lookup.test.js:34]: `Test` is not a defined function. Changed it to lowercase `test` to fix the error.
Screenshot 2025-07-19 at 22 08 04

File: Sprint-2/implement/querystring.js (hunk 3 above)
Comment (sedazam) [reply Sprint-2/implement/querystring.js:18]: | Character | Encoded As | Reason |
|-----------|------------|--------|
| `%`       | `%25`      | Starts an encoded sequence, must be escaped itself |
| ` ` (space) | `%20`    | Spaces are not allowed in URLs |
| `&`       | `%26`      | Separates query parameters |
| `=`       | `%3D`      | Assigns values to keys |
| `?`       | `%3F`      | Starts a query string |
| `!`       | `%21`      | Reserved character, often escaped |
| `/`       | `%2F`      | Used in paths, escaped in values |
| `#`       | `%23`      | Starts a URL fragment |
| `"`       | `%22`      | Double quotes must be escaped |
| `'`       | `%27`      | Single quotes can cause parsing issues |
| `:`       | `%3A`      | Used in protocols (e.g., `http:`) |
| `+`       | `%2B`      | May be interpreted as space in form data |

File: Sprint-2/implement/tally.test.js (hunk 4 above)
Comment (sedazam) [reply Sprint-2/implement/tally.test.js:26]: My original function used a plain object `{}`, which inherits built-in properties from `Object.prototype`, linked by default. This means it already has a `toString` method.

When I used `"toString"` as a key, it conflicted with the inherited method instead of behaving like a normal property.

To fix this, I used `Object.create(null)` to create a plain object that does **not** inherit from anything. This ensures that keys like `"toString"` can be used safely as regular properties.

File: Sprint-2/stretch/till.js (hunk 6 above)
Comment (sedazam) [reply Sprint-2/stretch/till.js]: Done

File: Sprint-2/stretch/till.js (hunk 7 above)
Comment (sedazam) [reply Sprint-2/stretch/till.js]: ### Why Use a Lookup Table for Coin Values

- Using a lookup table avoids fragile string parsing and makes the code more reliable.
- It improves readability and makes it easier to update or add new coin types.
- This approach ensures the function works consistently across different coin formats like `"1p"` and `"£1"`.

**Comment by sedazam** [comment]
[Merge branch 'main' into Sprint-2](https://github.com/CodeYourFuture/Module-Data-Groups/pull/578/commits/a2a063eedc1783750027deb339eb4a866765ce30)

What happened here?
